### DISCUSSION TITLE
gcc/clang optimisation tips for faster performance?

### DISCUSSION THREAD

**Discussion by al-caveman** [jedisct1/libsodium#1094] [Q&A]
Trying to reduce my app's runtime (critical in my case).  I thought of:

- Inlining libsodium's functions into my app.

Questions:

- Any idea how to achieve this function inlining?
- Any other tips?

**Comment by jedisct1** [comment]
You can compile with `--enable-opt` to get some minor additional optimizations, or compile with `zig cc` instead of `clang`.

> Any idea how to achieve this function inlining?

Enabling LTO (`-flto`) would do this automatically.

**Reply by al-caveman** [reply]
How safe is that?

I read this [here](https://libsodium.gitbook.io/doc/installation):
> Since different files are compiled for different CPU classes, and to prevent unwanted optimizations, link-time optimization (LTO) should not be used.

**Reply by jedisct1** [reply]
Unpredictable. Depends on what optimizations the compiler decides to do.

**Comment by covert-encryption** [comment]
For some reason chacha20-poly1305 runs much slower using the C interface than from Python (pynacl, which builds its own libsodium based on an older release). Encrypting a gigabyte of data (single message) takes 0.8 seconds from Python and 2.1 seconds on C API. I also tried compiling the git version and it runs just as slow as the build provided by my Linux distro. Decryption is equally slower compared to Python. Running on Ryzen 5.

Are the AVX/SSE optimisations enabled in default build, or do I need to specify something to get them? Any other reason for such big difference?

**Reply by jedisct1** [reply]
@covert-encryption are you properly initializing the library with `sodium_init()` prior to calling any other functions?

**Reply by covert-encryption** [reply]
Ah, that was it. A pretty stealthy bug :/

Now it runs in 0.6 seconds in C. I suppose it could be made even many times faster by using threads but this is a great improvement already.
